File: backend/app/api/care.py
"""Care metrics (the M1–M18 / C1–C6 report objects) and the raw observation
view behind them.

Both are read paths: ``care-metrics`` is the stored bundle slice behind
``ensure_fresh_assessment`` (no recompute beyond the staleness check) and
``raw-data`` is one grouped query over the observation table.
"""

from __future__ import annotations

from datetime import date, timedelta

from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy import func, select
from sqlalchemy.orm import Session

from app.api.worklist import ensure_fresh_assessment
from app.database import get_db
from app.engine.care.catalog import families_payload
from app.engine.care.pathways import pathway_by_key, pathway_for
from app.models.adherence import AdherenceTask
from app.models.checkin import Checkin
from app.models.observation import Observation
from app.models.patient import Patient

router = APIRouter(prefix="/patients", tags=["care"])

RAW_ROW_CAP = 400


def _get_patient(db: Session, patient_id: str) -> Patient:
    patient = db.get(Patient, patient_id)
    if patient is None:
        raise HTTPException(status_code=404, detail=f"Unknown patient: {patient_id}")
    return patient


@router.get("/{patient_id}/care-metrics")
def care_metrics(patient_id: str, db: Session = Depends(get_db)) -> dict:
    patient = _get_patient(db, patient_id)
    assessment = ensure_fresh_assessment(db, patient_id)
    care = (assessment.analytics or {}).get("care_metrics") or {}
    pathway = pathway_by_key(care.get("pathway")) if care.get("pathway") else pathway_for(patient)
    return {
        "version": care.get("version", "care-1"),
        "pathway": {"key": pathway.key, "name": pathway.name, "domain": pathway.domain},
        "headline": care.get("headline", []),
        "metrics": care.get("metrics", []),
        "families": families_payload(),
        "computed_at": assessment.computed_at.isoformat(),
    }


@router.get("/{patient_id}/raw-data")
def raw_data(
    patient_id: str,
    days: int = Query(30, ge=1, le=365),
    db: Session = Depends(get_db),
) -> dict:
    """Every non-deleted observation in the window, grouped by metric type,
    newest first, capped per type. The window is on ``local_date`` — the
    product's one day definition — not on ``start_time``."""
    _get_patient(db, patient_id)
    since = date.today() - timedelta(days=days)
    rows = db.scalars(
        select(Observation)
        .where(
            Observation.patient_id == patient_id,
            Observation.deleted_at.is_(None),
            Observation.local_date >= since,
        )
        .order_by(Observation.metric_type, Observation.local_date.desc(),
                  Observation.start_time.desc(), Observation.id.desc())
    ).all()

    grouped: dict[str, list[dict]] = {}
    summary: dict[str, dict] = {}
    for row in rows:
        key = str(row.metric_type)
        meta = summary.setdefault(key, {
            "metric_type": key, "unit": row.unit, "count": 0, "first": None, "last": None,
            "source_providers": [], "truncated": False,
        })
        meta["count"] += 1
        day = row.local_date.isoformat()
        meta["first"] = day if meta["first"] is None or day < meta["first"] else meta["first"]
        meta["last"] = day if meta["last"] is None or day > meta["last"] else meta["last"]
        provider = str(row.source_provider)
        if provider not in meta["source_providers"]:
            meta["source_providers"].append(provider)
        bucket = grouped.setdefault(key, [])
        if len(bucket) >= RAW_ROW_CAP:
            meta["truncated"] = True
            continue
        bucket.append({
            "date": day,
            "start_time": row.start_time.isoformat(),
            "value": row.value_num,
            "unit": row.unit,
            "source": provider,
            "granularity": str(row.granularity),
            "patient_reported": bool(row.is_patient_reported),
            "json": row.value_json,
        })

    checkins_count = db.scalar(
        select(func.count(Checkin.id)).where(Checkin.patient_id == patient_id)
    )
    tasks_count = db.scalar(
        select(func.count(AdherenceTask.id)).where(AdherenceTask.patient_id == patient_id)
    )
    return {
        "days": days,
        "since": since.isoformat(),
        "metric_types": sorted(summary.values(), key=lambda m: m["metric_type"]),
        "rows": grouped,
        "checkins_count": int(checkins_count or 0),
        "tasks_count": int(tasks_count or 0),
    }

File: frontend/src/api/care.ts
import { useQuery } from '@tanstack/react-query'
import { fetchJson } from './client'
import type { ConfidenceLevel, MetricStatus } from '../lib/risk'

/** Care-metrics contract (`GET /api/patients/{id}/care-metrics`): the stored
 *  analytics bundle slice for one patient — the pathway's headline picks plus
 *  every catalog metric, each a `CareMetric.to_dict()` from the engine. Kept
 *  apart from `queries.ts`/`types.ts` on purpose: another session owns those
 *  files while this page is being built. */

export type ChartKind = 'line' | 'band' | 'bars' | 'scatter' | 'dual' | 'gauge' | 'heat'

/** One chart point. Which keys are set depends on `ChartSpec.kind`:
 *  line/band `{x, y}`; bars/scatter `{x, y, label?}`; dual `{x, y, y2}`;
 *  heat `{x (day), row (task title), v (0 | 0.5 | 1 | null)}`; gauge has no
 *  points at all (its value lives in `extra`). */
export interface ChartPoint {
  x: number | string
  y?: number | null
  y2?: number | null
  label?: string
  row?: string
  v?: number | null
}

export interface ChartBand {
  x: number
  lo: number
  hi: number
}

export interface GaugeExtra {
  value: number
  min: number
  max: number
  bands: { to: number; tone: 'low' | 'med' | 'high' }[]
}

export interface ChartSpec {
  kind: ChartKind
  series: ChartPoint[]
  reference: number | null
  band: ChartBand[] | null
  x_label: string
  y_label: string
  y2_label: string
  marker_x: number | null
  fit: { x: number; y: number }[] | null
  extra: Record<string, unknown>
}

/** Component rows behind M12/M13/M14/M15/M16/M17/M18/C5. Shape varies by
 *  metric: most carry `label`; M14's rows are per-task (`title`, `task_id`,
 *  `rate`, `verified`, `self_attested`, `missed`). Not rendered yet. */
export interface MetricDriver {
  metric_type?: string
  label?: string
  title?: string
  contribution?: number
  direction?: string
  [key: string]: unknown
}

export interface CareMetric {
  id: string
  key: string
  name: string
  family: string
  template: string
  feasibility: string
  tier: number
  status: MetricStatus
  status_text: string
  value: string | null
  value_num: number | null
  unit: string
  value_label: string
  delta_text: string | null
  finding: string
  next_step: string | null
  confidence: ConfidenceLevel
  coverage_text: string
  guarded: boolean
  chart: ChartSpec | null
  method: string
  inputs: string[]
  unlock: string | null
  feeds_from_tasks: string[]
  drivers: MetricDriver[]
  /** False when the metric is not meaningful on this patient's pathway —
   *  the Full stats view folds those into a collapsed tail. */
  applicable: boolean
  domains: string[]
}

export interface CareFamily {
  key: string
  name: string
  metric_ids: string[]
}

export interface CarePathway {
  key: string
  name: string
  domain: string
}

export interface CareMetricsResponse {
  version: string
  pathway: CarePathway
  /** Metric ids in display order — the API already picks applicable ones. */
  headline: string[]
  metrics: CareMetric[]
  families: CareFamily[]
  computed_at: string
}

export interface RawDataRow {
  date: string
  start_time: string | null
  value: number | null
  unit: string
  source: string
  granularity: string
  patient_reported: boolean
  json: Record<string, unknown> | null
}

export interface RawDataType {
  metric_type: string
  unit: string
  count: number
  first: string | null
  last: string | null
  source_providers: string[]
  /** Set when this type hit the 400-row cap for the window. */
  truncated?: boolean
}

export interface RawDataResponse {
  days?: number
  metric_types: RawDataType[]
  rows: Record<string, RawDataRow[]>
  /** Some servers report the cap once for the whole payload. */
  truncated?: boolean
  checkins_count: number
  tasks_count: number
}

/** Eager on purpose: the headline tiles sit above the fold. The key lives
 *  under the `['patient', id]` prefix so a Refresh analysis sweeps it up with
 *  the rest of the page (`recomputeKeys`). */
export function useCareMetrics(id: string) {
  return useQuery({
    queryKey: ['patient', id, 'care-metrics'],
    queryFn: () => fetchJson<CareMetricsResponse>(`/api/patients/${id}/care-metrics`),
  })
}

/** Lazy: only the Raw data tab pays for the observation dump. */
export function useRawData(id: string, days: number, enabled = true) {
  return useQuery({
    queryKey: ['patient', id, 'raw-data', days],
    queryFn: () => fetchJson<RawDataResponse>(`/api/patients/${id}/raw-data?days=${days}`),
    enabled,
  })
}
